### ISSUE TITLE
Decision Proposal 288 - Non-Functional Requirements Revision

### ISSUE THREAD

**Issue by CDR-API-Stream** [ConsumerDataStandardsAustralia/standards#288]
This decision proposal contains a proposal for changes to Non-Functional Requirements and the Get Metrics end point based on feedback received through various channels.

The decision proposal is embedded below:
[Decision Proposal 288 - Non-Functional Requirements Revision.pdf](https://github.com/ConsumerDataStandardsAustralia/standards/files/10830330/Decision.Proposal.288.-.Non-Functional.Requirements.Revision.pdf)

~~Consultation on this proposal will close on the 7th April 2023.~~

Note that the proposed changes have been republished in [this comment](https://github.com/ConsumerDataStandardsAustralia/standards/issues/288#issuecomment-1515771012) and [this comment](https://github.com/ConsumerDataStandardsAustralia/standards/issues/288#issuecomment-1534077983)

Consultation will be extended to obtain feedback on these updated changes until the ~~5th May 2023~~ ~~12th May 2023~~ 19th May 2023.

**Comment by CDR-API-Stream** [comment]
The [opening comment](https://github.com/ConsumerDataStandardsAustralia/standards/issues/288#issue-1524928359) has been updated and the Decision Proposal is now attached.  

The Data Standards Body welcomes your feedback.

**Comment by ranjankg2000** [comment]
In the recommended “consent metrics”, we can potentially look at splitting authorisation count for accounts between individual and non- individual entity. It can provide an indication of adoption among businesses

**Comment by AusBanking-Christos** [comment]
On behalf of the ABA, can we please request an extra week for our Members to respond.  Thank you.

**Comment by CDR-API-Stream** [comment]
Of course, no problem.  We'll extend the consultation until 7th April

**Comment by damircuca** [comment]
We recommend making the “GET /admin/metrics” endpoint publicly accessible without any authentication or protection. This change would provide numerous benefits to the ecosystem.

`Ref: https://consumerdatastandardsaustralia.github.io/standards/#get-metrics`

- It would increase transparency by allowing stakeholders to see what is working and what is not. Currently, the only way to determine whether a data holder is operational is by encountering errors or checking the CDR website that hosts a status report. With public access to metrics, stakeholders can easily access this information and make informed decisions.

- Organisations would be able to incorporate these statistics into their consent flow, which would improve the consumer experience. By querying the API, organisations can determine which data holders are operational and avoid presenting options that may result in errors or unsuccessful experiences for the consumer.

- Public access to metrics would encourage the development of tooling and reports that non-developers can use when working within the ecosystem. This would empower first line support personnel to field calls about potential issues with individual data holders and improve overall support within the ecosystem.

We believe that restricting access to the “GET /admin/metrics” endpoint only to the ACCC and individual data holders limits the potential benefits to the ecosystem. By allowing public access, ADRs and other stakeholders can make better-informed decisions and plan their approach to each data holder more effectively.

**Comment by nils-work** [comment]
Hi @damircuca 

Are you finding cases where the [Get Status](https://consumerdatastandardsaustralia.github.io/standards/#get-status) endpoint does not accurately represent implementation availability (i.e., because you encounter unexpected unavailability), or that there is not enough detail (on specific endpoint availability, for example) for it to be useful when initiating a series of requests?

**Comment by jimbasiq** [comment]
> Hi @damircuca
> 
> Are you finding cases where the [Get Status](https://consumerdatastandardsaustralia.github.io/standards/#get-status) endpoint does not accurately represent implementation availability (i.e., because you encounter unexpected unavailability), or that there is not enough detail (on specific endpoint availability, for example) for it to be useful when initiating a series of requests?

Hi @nils-work 

If you are able, please take a look at https://cdrservicemanagement.atlassian.net/servicedesk/customer/portal/2/CDR-3328 to see an example of when the [Get Status](https://consumerdatastandardsaustralia.github.io/standards/#get-status) endpoint has not worked sufficiently.

You can see the failure reflected in the CDR performance dashboard screengrab below. Availability is apparently 100% but a 50% drop in API traffic is visible i.e. APIs are down. Specifically Error 500s on data retrieval APIs.

![image](https://user-images.githubusercontent.com/86987893/227416473-bebde5a5-9188-475b-96a6-33b508b7a3fc.png)

**Comment by nils-work** [comment]
Thanks @jimbasiq, I'm not sure if I'll be able to access that ticket, but I'll check.

As a general comment, and it may not have been the case, but my initial thinking is that a scheduled outage may produce this effect (I note the drop in traffic appears to be over a weekend).

The Availability metric (at ~100%) would not be affected by a scheduled outage, but any invocations (resulting in 500s) may still be recorded and reported in Metrics (though there is not an expectation of this during an outage).

This makes it appear that either the Status `SCHEDULED_OUTAGE` was ignored by clients and about 50% of the average invocations were still being received (perhaps only some endpoints were affected), or the status was incorrectly reported as `OK` during an unexpected outage (of about 3 days) but only about 50% of invocations could actually be logged. 

If it was an unexpected outage, the Status response should have been `PARTIAL_FAILURE` or `UNAVAILABLE` and Availability should have been about 90% for the month.

**Comment by damircuca** [comment]
Hey @nils-work generally the more data that is available the more options we have on how to incorporate it within the delivery of CDR services. 

The Get Status endpoint is very coarse, and doesn't provide enough depth for us. Whereas metrics has a lot more detail that can be used to better support customers, implement a more refined CDR CX flow, and also help with fault resolution. 

Even now, whenever a support ticket is raised, we find ourselves going straight to the metrics report (available via CDR site) to see what the lay of the land looks like before we respond back to our customers.

Further to this, even with scraping we have bee surfacing metrics such as performance metrics, stability percentages and more, which we find valuable to drive decisions and future enhancements.

Realise it may be a big ask, but it will be valuable - and will also raise the transparency and accountability within the ecosystem which is equally important for making a success.

**Comment by CDRPhoenix** [comment]
Rather than opening up the Get Metric endpoint to the public, I think it is worthwhile to allow the public to sign up and download raw data from Get Metrics that ACCC CDR collects, this will still place the responsibility of chasing non responding Data Holder Brands with CDR and wont flood the Data Holders brands with Get Metrics request.

As a side note stemming out of ADRs hitting Maximum TPS tresholds, I think it is also worthwhile to revisit batch CDR requests and whether there is a need for something like Get Bulk Transaction for banking if there is a use case for get fetch transaction periodically and for DSB to consider creating a best practice article on ZenDesk on ADR calling patterns, i.e. do you need to perform a get Customer Detail, get Account details every time you want to pull transactions down. Otherwise any increase in traffic threshold will be soaked up with "low value" calls and we will be forever chasing for more and more bandwidth.

**Comment by CDR-API-Stream** [comment]
In response to @ranjankg2000:

> In the recommended “consent metrics”, we can potentially look at splitting authorisation count for accounts between individual and non- individual entity. It can provide an indication of adoption among businesses

Thank you for this feedback.  This is a good idea to incorporate

**Comment by CDR-API-Stream** [comment]
In response to @damircuca:

Making the metrics API public is a very interesting idea.  The DSB will discuss this internally with Treasury and the ACCC to identify any policy reasons why this would not be possible.  There are real no technical reasons why this would be an issue provided there were low non-functional requirements that would ensure Data Holder implementations didn't need to be over-scaled.

**Comment by damircuca** [comment]
One thing to consider which CDRPheonix touched on was to open up the data that ACCC collects vs forcing the Data Holders to make changes on their end. Sorry for stating the obvious your likely considering this already 🤷🏻‍♂️

**Comment by ACCC-CDR** [comment]
The ACCC supports the changes outlined in Decision Proposal 288. These changes will improve the accuracy of the information captured through Get Metrics and better support the estimation of consumer uptake.

The ACCC suggests a further change to the PerformanceMetrics value. Currently, it is proposed that this value be split into unathenticated and authenticated metrics. The ACCC suggests that splitting this value by performance tier (i.e. Unauthenticated, High Priority, Low Priority, Unattended etc.) would better align these measures with the metrics reported for invocations and averageResponse. This change would assist the ACCC’s monitoring of Data Holder’s compliance with the performance requirements.

The ACCC notes suggestions by participants regarding the availability of Get Metrics data. As flagged by the DSB above, the ACCC will continue to collaborate with its regulatory partners to assess how Get Metrics data can most effectively enhance the CDR ecosystem but suggests that such measures should be considered separately from this decision.

**Comment by cuctrangsb** [comment]
Overall this will be a large-sized change for Great Southern Bank to implement. Given we have already planned work up until July 2023, it would be much appreciated if the obligation date for this change can be at least 6 months once the decision is made.

**Issue: Error code mappings.**
Out of the 2 options proposed, we prefer option 1 - Split the error counter to http error code and corresponding counts. This will give better underrating on what all error codes application retuned. Currently all 5XX considering as same.

**Issue: Lack of consent metrics**
**We need clarification on historical records - are we looking for counts (authorisation/revocation) from beginning or the last 8 days?** All other metrics data carrying last 8 days' data, but customer count and recipient count is not.

**Issue: Scaling for large data holders.**
We prefer Option 2 - tiered TPS and Session count NFRs based on the number of active authorisations the data holder has.
The customer base and the uptake of Open Banking at Great Southern Bank remain relatively small compared to the big banks. This option will help us reduce the cost of maintaining our infrastructure to meet the current TPS requirements. We can proactively manage the active authorisation and scale up slowly as required. Depending on the tiered threshold, potentially we can look at ensuring we meet the current tier plus the next tier up to cater for any sudden influx of registrations.
Further consultation to define the tiered threshold will be much appreciated.

**Comment by anzbankau** [comment]
We are broadly supportive of proposal to uplift the CDR’s non-functional requirements as outlined in Decision proposal 288. This decision proposal describes a range of topics, and we suggest any proposed implementation schedule be priority driven with careful consideration given to improved consumer outcomes, ecosystem value and impact to data holders (i.e., cost, time, and complexity of implementation).
 
Specific points of feedback as follows:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lucasj1/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lucasj1/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Item | Feedback
-- | --
Get Metrics Issues:   authenticated and unauthenticated | We support the recommendation to "Split between   authenticated and unauthenticated".
Get Metrics   Issues: Granularity alignment to NFRs | We are not clear on the benefit, consumer or otherwise, for   reporting on an hourly basis. Suggest that the benefit is clearly articulated   before changes are made.
Get   Metrics Issues: Lack of consent metrics | As above we suggest that   the consumer benefit of this proposal be clearly articulated as the effort to   implement is likely to be significant.
NFR Issues: Scaling for   large data holders | We support scaling at   brand level and the uplift of the TPS ceiling in accordance with appropriate   demand forecast. We have yet to see evidence that the current TPS ceiling is   inadequate and/or adversely affecting consumer outcomes. We suggest an   evidence-based approach to forecasting future demand, so that holders can   plan implementations with sufficient lead time. We note that open banking   digital channels are not like-for-like with existing banking channels, the   CDR rules for entitlements mean that there are additional performance   considerations for data holders.          We welcome the opportunity to work with the DSB to review the current   demand on the system.          We do not support removing site wide NFRs.
NFR Issues: NFRs for   incident response | We are broadly supportive   of NFRs for incident response and endorse a more transparent approach to   tracking issues and resolution.          NFRs for incident resolution are problematic as there is no easy way to   guarantee resolution times, particularly with complex issues which require   interactions between ADRs, consumers and data holders; these interactions can   be laboured and transactional, owing to the limited information which can be   exchanged outside of the CDR.           We are also unclear how issues can be objectively and consistently   classified in terms of severity and prioritisation without independent   mediation.
Implementation   Considerations | Per earlier point,   implementation should be priority-based with appropriate consideration given   to the ecosystem’s capacity for change and demonstrable consumer benefit. A   more predictable change cadence with sufficient lead time for implementation   is recommended.
Get   Metrics changes:Errors | We recommend to count by   HTTP status code rather than URN.



</body>

</html>

**Comment by kristyTPGT** [comment]
TPGT appreciates the opportunity to provide feedback in relation to Decision Proposal 288. Please find our feedback attached.
[DP-288 Final Response.pdf](https://github.com/ConsumerDataStandardsAustralia/standards/files/11155664/DP-288.Final.Response.pdf)

**Comment by johnAEMO** [comment]
AEMO thanks you for the opportunity to respond to this Decision Proposal

In terms of feedback to the getMetrics API, AEMO has the following comments:

1. The definitions of each field could be made clearer by including the httpStatus codes applicable to each field.  That is:
•	Availability – is presumed to be the % all requests not returning a 5xx series status code
•	Performance – is presumed to be for all successful requests (200s) within thresholds
•	Invocations – is presumed to be all requests (successful or not)
•	Average response – is presumed to be for all successful requests (200s)
•	Tps fields – is presumed to be all requests (successful or not)
•	Errors – are presumed to be all 5xx series status code responses
•	Rejections – are presumed to be all 429 series status code responses (traffic threshold limits)
Missing:
•	Other 400 series errors should be reported. These are not currently reported in getMetrics and in some instances are significant in number and complete the overall picture of request quality
•	95th percentile – is statistically a more useful indicator of overall performance spread than the currently requested mean

2.	NFRs for AEMO as a secondary data holder who is without access to the necessary fields to determine if the request is Customer Present or Not Present.  At this stage we have made assumptions that the customer is present, except where there are multiple service points requested (getServicePoints API only). 

3.	Performance Observations
AEMO currently has issues with its Usage APIs’ performance in providing large payloads and is undertaking a proof of concept to identify where and how to best address this.  There are two changes in the industry that will increase payment size in the short term and medium term. Both changes will likely impact performance.
•	In the short term the tranche 2 obligation for complex requests is expected to include additional multiple service points in one request and that will impact performance. 
•	In the medium term, the industry is planning to accelerate the upgrade from basic meters to interval meters as part of the AEMCs ‘Review of the regulatory framework for metering services’. The objective of the review is to replace all basic meters with 5 minute interval meters in the NEM; each meter will provide 288 interval reads per day per unit of measure. 

While we accept that AEMO is obliged to service every request it receives, there are some observations we have already made that may improve the ADRs’ experience of this service:
•	AEMO receives interval meter Usage data from meter data providers at best the day after the reading is taken – multiple API requests within a day will not yield any more up-to-date data.  Interval meter readings of 5-30 minutes are used by the energy industry to settle the market and to charge retailers for the energy their consumers have used during each interval, (who in turn use this to bill their consumers).  While the meter might read every 5-30 minutes, this does not indicate the frequency that usage data is circulated across the industry.
•	AEMO Basic meter Usage data is typically read on a 1-3 monthly basis and it too is shared across the energy industry at best the day after reading.  Similarly, multiple requests in a day will not yield more up-to-date data.

**Comment by AusBanking-Christos** [comment]
6 April 2023

**Submitted on Thursday, 6 April 2023 via:  [Consumer Data Standards Australia - GitHub site]**(https://github.com/ConsumerDataStandardsAustralia/standards/issues/288) 

Dear [@CDR-API-Stream](https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCDR-API-Stream&data=05%7C01%7CJulian.Luton%40cba.com.au%7C7cc4f852657d43522a0108db34a6fb53%7Cdddffba06c174f3497483fa5e08cc366%7C0%7C0%7C638161665657899499%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=buKD3iX%2BsiygVan5b9MpufesDPB%2FSOlOzHxp%2ByD7BUk%3D&reserved=0)

**ABA Response to Decision Proposal 288 – Non-Functional Requirements**


The Australian Banking Association (**ABA**) welcomes the opportunity to respond on behalf of our Members regarding DP 288 [Non-Functional Requirements.](https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FConsumerDataStandardsAustralia%2Fstandards%2Fissues%2F288&data=05%7C01%7CJulian.Luton%40cba.com.au%7C7cc4f852657d43522a0108db34a6fb53%7Cdddffba06c174f3497483fa5e08cc366%7C0%7C0%7C638161665657899499%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=MHiFVf4FIt9pZxE8IFdJi%2Fnkon6ZtlkRfDAxff3n7hk%3D&reserved=0)  

The ABA has met with Members to discuss [DP 288](https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FConsumerDataStandardsAustralia%2Fstandards%2Ffiles%2F10830330%2FDecision.Proposal.288.-.Non-Functional.Requirements.Revision.pdf&data=05%7C01%7CJulian.Luton%40cba.com.au%7C7cc4f852657d43522a0108db34a6fb53%7Cdddffba06c174f3497483fa5e08cc366%7C0%7C0%7C638161665657899499%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=zAwWKSxqNhcddDO5ITiCU6wX9CqX44pPssaA2Wm50JI%3D&reserved=0) in more detail and provides the following feedback. 

A point raised by Members centred on [Dynamic Client Registration (DCR) Response Time - NFR.](https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FConsumerDataStandardsAustralia%2Fstandards-maintenance%2Fissues%2F409%23issuecomment-1215969464&data=05%7C01%7CJulian.Luton%40cba.com.au%7C7cc4f852657d43522a0108db34a6fb53%7Cdddffba06c174f3497483fa5e08cc366%7C0%7C0%7C638161665657899499%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=QRd7vOLD6Q1clSQF4aoStmsMbPo8ez1Ov9NENaqpLxk%3D&reserved=0) Members noted that current CD Standards (**the Standards**) response times for DCR can prove challenging to comply with, given the additional latency taken by the Accredited Data Recipient’s (**ADRs**) to undertake the registration request JWT validation required by the Standards. This additional latency time which Data Holders (**DH**) rely on, in relation ADRs to **outbound connection whitelisting** is not split out from times noted in the Standards. We propose that the DSB further reconsiders amending response times to reflect this. As a point of reference, we include a link to last year’s [Dynamic Client Registration Response Time NFR #409.](https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FConsumerDataStandardsAustralia%2Fstandards-maintenance%2Fissues%2F409%23issuecomment-1215969464&data=05%7C01%7CJulian.Luton%40cba.com.au%7C7cc4f852657d43522a0108db34a6fb53%7Cdddffba06c174f3497483fa5e08cc366%7C0%7C0%7C638161665658056090%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=UzJrMQ%2BDuid6ymPy%2Bypjey4j7LblZ73qIZET0xe8pqk%3D&reserved=0) 

We note DP 288 confirms that the DCR will not be subject to change but reserved for a future direction of the **Register Standards**. We ask the DSB to reconsider our Members’ position to address their concerns on this point as part of the development requirements in the DP 288. 

Equally, we do thank the DSB for providing clarity on the origins of the six new Consent Metrics (new authorisations) introduced into DP 288. Where the DSB explained that the ACCC requested these specific new Consent Metrics, so the ACCC can determine where customers are **dropping off in the consent flow**, and whether this is occurring at the ADR or DH ends. 

We propose an open discussion or workshop with the ACCC regarding their request for additional consent metrics as a way to understand and improve consent drop off rates. The cost and effort to add these metrics, when aggregated across all DHs is significant. We propose that a small number of DHs that between them cover most consent flow types (basically looking at different OTP delivery mechanisms), to volunteer and provide the metrics requested on a one-off basis as input for a study into improvement in consent flow UX, which is presumably what the ACCC want the metrics for in the first place. This would lead to a faster outcome and be cheaper for not only all DHs but also for the volunteers (as they would not be extending the Metrics API, only collating the data on a one-off basis). We also note that consent flow is likely to change radically because of Action Initiation and the introduction of FAPI 2.0 and RAR.

Should the above volunteer proposal not be accepted, some Members have raised comment on the DP 288 section around **Implementation Considerations**, which includes the six new Consent Metrics. We note that the DSB acknowledged that it was prepared to **ramp up the implementation schedule over an extended period**. An initial proposal raised by the DSB was for five years being a potential period of implementation. The ABA welcomes this proposal by the DSB, to allow our Members to better resource and budget accordingly for these, and other priorities, including those planned for future CDR implementation (e.g., Action Initiation and Non-Bank Lending).

We would also ask the DSB further considers how it would prefer to update the CX journey negative path. At any point along the customer journey, the customer can decide to cancel and there can be multiple reasons why (regardless of whether the customer is still at ADR side or DH side). Recognising it is not only about the customer hitting a technical issue and can’t continue with the journey. Ideally at the point of customer-initiated cancellation, data should be collected as to why the customer decides to cancel and it should be a “standardised” set of reasons Members can all report on. Currently (incl., ADR/DH) this data is not collected from customer when they cancel, as it is deemed as introducing friction when it is not on the DSB’s CX flow.

We generally understand the DSB’s proposal to balance the requirement on TPS ceiling obligations tied to the **number of consents** held by each bank. Meaning this is intended by the DSB to be a fairer allocation of investment across individual banks. As opposed to setting a fixed figure which some smaller Members may result in excessive systems costs based on TPS ceiling measure which banks are not likely to reach. 

Members have expressed challenges on TPS thresholds around provisioning for peak times. Members have suggested further workshops be facilitated by the ACCC and DSB on how to address this matter on TPS and response time concerns and achieve a fair and reasonable model across all industries and emerging areas like Action Initiation. Members believed this approach could better serve reaching a resolution than direct feedback to a DP. 

We would rather have a staged lift in TPS that is tied to a realistic industry consensus forecast. If the increase is staged over a number of years, we would also like a mechanism to periodically revise the required TPS as more data becomes available. Alternatively, if a formulaic approach tied to consents is taken, we would expect that the formula be deployed in a manner that gives Members enough time to budget for and implement system uplifts to cater for increased TPS NFRs, including systems changes for third party service providers. 

We also propose that demand management is considered. For example, demand from ADRs could be spread across 24 hours, and not 3 hours in the early morning. This could be enforced through hourly quotas. Another consideration is restricting the number of times that slow moving data is queried. If a given data set is only updated daily, then this could be flagged with a new metadata field that the ADRs would have to respect and only request that data once a day.

In conclusion, we note DP 288 raises challenges for smaller DHs around TPS and consents, with a few options raised by the DSB to remediate under the heading, **Scaling for large DHs**. One proposal includes, ‘**increase in the site wide TPS and Session Count NFRs**’. Some Members have requested evidenced based data that the DSB sees why, or foresees in the ecosystem to warrant a change, and the types of changes being proposed by the DSB.  

Further discussions or workshops with the ACCC and the DSB to discuss these NFRs and other appropriate matters to understand how this potential proposal could be applied efficiently, would benefit our Members. Given if it were applied by the DSB, to accommodate for a rise in TPS ceiling thresholds, this would likely result in significant investment to affected ABA Members. 

We thank the DSB again for the opportunity to respond on behalf of our Members, as we are equally thankful for the DSB extending our response date by a week.

We look forward to continuing our engagement and thank the DSB for its support in these matters.

Yours sincerely

Australian Banking Association

**Comment by Telstra-CDR** [comment]
Please find attached feedback from Telstra
[DP288 - Feedback.pdf](https://github.com/ConsumerDataStandardsAustralia/standards/files/11185435/DP288.-.Feedback.pdf)

**Comment by CDR-API-Stream** [comment]
Thanks everyone for all of the feedback.  There was a lot that came in just before or over the Easter weekend.  We are going through the feedback and will respond incrementally over the next couple of days.

We will leave this consultation open during this time and for a further couple of days so that everyone can respond to what we will be proposing to take to the chair.

**Comment by jimbasiq** [comment]
Considering the comments around the difficulty of Data Holder implementation whilst balancing other work and obligations, Basiq would be supportive of a phased delivery approach and further discussion is required to agree and prioritise "most useful" and "easier to implement" metrics. I would prefer to have several most useful metrics in three months rather than all metrics in 12 months.

**Comment by jimbasiq** [comment]
On the topic of a TPS metric. It is always going to be a challenge for Data Holders to "right size" their infrastructure in order to avoid negatively affecting a consumer. For instance crystal balls or true elastic scalability will be required to set TPS and Session Count NFRs based on the number of active authorisations the data holder has.

Can I suggest the TPS metrics drives the ongoing obligation. i.e. Data Holders do not just report on TPS but on % utilisation of their current limit. If metrics show TPS is regularly exceeding a defined threshold (e.g 90%) the Data Holder should be obligated to raise their TPS.

**Comment by jimbasiq** [comment]
One last comment on the Consent Metrics.

For _abandonedConsentCount
The number of initiated consent authorisation flows abandoned (for any reason) during the period, reported on a daily basis_

Could we get more granular than "for any reason", a Data Holder should be able to detect the different between:

- abandon without interaction from the consumer
- failure to successfully authenticate
- stage at which the consumer abandoned post authentication

**Comment by CDR-API-Stream** [comment]
Please note the attached feedback provided via email by AGL:
[AGL - Non-Functional Requirements Revision.pdf](https://github.com/ConsumerDataStandardsAustralia/standards/files/11243805/AGL.-.Non-Functional.Requirements.Revision.pdf)

**Comment by CDR-API-Stream** [comment]
Thanks everyone for the volume of feedback but also for the depth of analysis and insight contained in the feedback.  It has taken a while to parse all of the feedback due to the quality and some of the initial assumptions of the DSB as to a likely proposal to the chair have changed as a consequence.

For this reason the DSB has decided to do the following:
1. Summarise our understanding and interpretation of the feedback received on each of the key topics in the consultation
2. Present a proposal of the changes we intend to recommend to the Chair based on the feedback
3. Leave the consultation open for a further two weeks to let the community to respond to our proposal to clarify any misunderstandings we may have to improve the proposed changes

**Comment by CDR-API-Stream** [comment]
Summary of feedback received on the key issues in the consultation:

## Changes to existing Metrics

* There was minimal commentary on the changes to the Get Metrics API to make the field types more specific so we are assuming that this a non-controversial change.
* There was consensus support from those that commented that error types should be reported at the HTTP error code level.
* There was support for splitting availability and TPS metrics between authenticated and unauthenticated instances.
* There was also support for splitting performance metrics between authenticated and unauthenticated but also adding a split for the various performance tiering levels for authenticated APIs.  As the primary consumer of this data for regulatory purposes the DSB weights this feedback heavily

We think we will be able to incorporate this feedback into the proposed changes to the standards without too much difficulty.

## New Metrics

* While implementation issues were raised the addition of consent metrics as requested by the regulator was generally supported
* The proposed consent metrics were generally supported with the following suggestions for change
 * It was suggested that consent counts be split between individual and non-individual customer types
 * The reasons or stage of authorisation failure would also be a good addition

We will be incorporating this feedback into the proposed changes noting that we need to maintain a balance between the usefulness of the data and the cost of implementation. 

Great Southern Bank suggested that it would be helpful to have clarity over how far back metrics need to be provided at go live of any changes.  This is helpful feedback and will be incorporated into the proposed changes.


## Changes to NFRs

The feedback on the non-functional requirements was extensive and raised some real issues that were not canvassed in the Decision Proposal.  General feedback will be responded to first with specific topics being addressed afterwards.

The ABA noted a proposal for a five year schedule that was discussed in a bilateral meeting that deserves to be more clearly articulated.  It is understood by the DSB that there is a long lead time for the planning for capital expenditure on infrastructure so a long term, predictable, plan for NFR changes would be very helpful.  The DSB therefore suggested that a multi-year plan for changes to NFR thresholds could be published.  This plan could then be reviewed annually in a workshop of participants and agencies.  This aligns with feedback received from the ADR community about the need for NFRs to constantly evolve with the ecosystem.  This approach will be considered in the proposed solution.

It was suggested by participants in the Telco and Energy sectors that there should be sectoral differences in NFRs.  The NFRs are designed to provide a consistent and predictable service to ADRs so that they can operationally plan and manage their solutions and are therefore not specifically aligned to the sector of the Data Holder.  Also, sector specific NFRs would become complex to manage and is not aligned with the cross-sectoral approach of the CDR data standards.  That said, the reasons for these sectoral differences were provided and align with the concerns raised in the Banking sector about the same NFRs being applied to small banks as to large banks.  For instance, it would appear that a very small ADI in the banking sector is likely to have less technical capacity than a large telco or energy retailer.  As a result these issues will be address via tiering for data type and number of authorisations rather than on a sectoral basis.

It was suggested by one participant that the existing NFRs were not extensively consulted on.  The DSB reject this position.  The existing NFRs were consulted on multiple times and many months of active operation was allowed before the NFRs became enforceable so that participants could communicate concerns related to the NFRs and request amendments.  The DSB were clear that this was the intended approach.  This consultation is itself another step in that ongoing consultation process.

There were a number of participants that called for using an evidence based approach to defining the NFRs.  This is the approach preferred by the DSB and we have been using all of the data we are able to obtain to develop the standards.  We are able to access the existing metrics data provided to the ACCC by Data Holders and a number of ADRs have been forthcoming with data through various consultation channels.  Data from Data Holders that they feel would be helpful in the Chair being able to make evidence based decisions would be extremely welcome.

### Tiering of TPS

There was consensus support for a tiered approach to TPS thresholds based on number of active authorisations.  In the absence of any specific data provided by participants the first proposal for such a tiering strategy will be based on existing metrics data.

The suggestion by Basiq that setting these tiers based on existing average and peak TPS outcomes reported by Data Holders is very useful in this regard.


### NFRs for Low Velocity Data

Multiple participants provided feedback that low velocity data sets (such as energy usage data from AEMO that only updates on a daily basis) are being requested far more often than necessary by ADRs.  This was echoed by the ABA for slow moving data sets in the banking sector.

This is very useful feedback and is currently addressed at a high resolution by the [NFRs applicable to ADRs](https://consumerdatastandardsaustralia.github.io/standards/#traffic-thresholds).

There is currently not mechanism for Data Holders to respond to breaches of these requirements by ADRs with a 429 error code. To address this the proposal will incorporate a tiering structure to indicate the number of duplicate requests for a specific resource that Data Holders are required to service before being permitted to expect the ADR to be caching the response provided.


### White Labelled Brands

One participant raised specific concerns around NFRs related to white labelled brands.  The DSB proposes to discuss this issues with the participant bilaterally to further understand the concerns raised.  First review of these concerns indicate, however, that the issues are specific to the implementation choices made inside the ADI and are not generalised issues faced by all white label solution providers.  As a result these issues will not be addressed in the proposed solution but further consultation will be undertaken.


### DCR NFRs

It was noted in the Decision Proposal that DCR specific NFRs would not be addressed in this consultation but would be raised in the consultation on the future of the Register Standards.  This consultation is now open at: [Noting Paper 289 - Register Standards Revision](https://github.com/ConsumerDataStandardsAustralia/standards/issues/289)


### NFRs for Incident Response

As noted in the Decision Proposal, requirements for specific response times for incidents will be considered separately via the existing incident management working group.  


## Implementation Considerations

There was competing feedback given for the implementation of the changes arising from this consultation.  On one hand, there was reasonable feedback provided concerning the cost and time of delivery.  On the other hand, there was an urgent need articulated for some of the issues to be addressed as soon as practical.

The DSB is also conscious of the need to provide clarity for the second tier energy retailers going live later this year.

It was suggested by Basiq that a phased approach be used to allow for some of the more important changes be incorporated faster than some of the less critical changes.

The proposed solution approach will take this feedback into account and seek to strike a balance between need and cost.


## Additional Feedback and Comments

It was suggested that clarification of terms in NFRs and metrics would be very helpful with those implementing data holder solutions.  The DSB agrees with this and has been weighing the benefit of expanding the detail in the standards, along with the associated increasing in complexity of the standards, and providing more expansive guidance.  We have opted to develop improved guidance to address this issue as it allows us to provide clarity more rapidly when it is needed.

It was suggested that metrics data should be made public so that anyone can call a Data Holder’s metrics API.  The rules allow that the standards related to metrics is for public reporting so it is understood that there is no regulatory barrier to this approach.  This would, however, be a big change and would mean that the Get Metric API will receive a lot more calls than originally intended.  It was also suggested that the ACCC could make the data they collect available to participants for an API which would address this concern.  While this is a very helpful suggestion and a good topic to be considered it was not the intended focus of this consultation and it has been requested by the ACCC that this be considered separately.  As a result a change of this nature will not be incorporated into changes resulting form this Decision Proposal.

It was raised by AEMO that they do not receive visibility of whether an invocation is customer present or not present so cannot scale accordingly.  This was an oversight as they are not required to report on NFRs but it is clear based on the feedback received that this would be helpful for AEMO in managing their infrastructure.  As a result a solution to address this will be raised as a CR to be considered in maintenance iteration 15.

The ABA and others suggested that data around the CX flow that would help improve how the authorisation process evolves would be helpful.  We would welcome such data.  This would best be handled via other consultations and workshops related to authentication uplift.

**Comment by CDR-API-Stream** [comment]
At the DSAC on Wednesday the 19th April a presentation was provided by a member on operational concerns with the CDR ecosystem.

In the discussion arising from this presentation it was identified that the ability to understand the stage at which a customer abandons the authorisation flow would be helpful in identifying CX problems with specific Data Holders.

We will attempt to incorporate this feedback into the changes to the metrics API.

**Comment by spikejump** [comment]
> At the DSAC on Wednesday the 19th April a presentation was provided by a member on operational concerns with the CDR ecosystem.

Is this presentation shareable publicly?  or was it a live presentation?

**Comment by CDR-API-Stream** [comment]
It was a live presentation.  The DSAC member is free to publish the presentation if they wish but the practice of the DSAC is not to attribute contributions to specific members.  If it is made public it will be via the minutes of the DSAC which take a while to process.

**Comment by CDR-API-Stream** [comment]
Here are the proposed changes to Get Metrics for further feedback.  These are candidate changes to be proposed to the Chair unless there is feedback indicating they should change:

## Proposed Metrics Changes

There will be two additional versions of the Get Metrics API defined - v4 and v5 - with different obligation dates.

V4 will contain additions that are considered less impactful for implementors and will be targeted to align with the obligation dates for the remaining energy data holders in the later half of 2023.

V5 will contain additional metrics that may require more instrumentation to be able to achieve and will be targeted for Q1 2024.

As there is only a single client for the Get Metrics API data holders will not be required to maintain older versions of the Get Metrics API but they must pay attention to the fact that the ACCC, when calling this API, will likely call with an `x-v` value of 5 and an `x-min-v` value of 3 and implementations should successfully respond with their highest supported version based on the current requirements of the standards.

### V4 - Obligation Date Y23 #5 (13/11/2023)

Type changes:
* The type of availability and performance metrics will be changed to `RateString` and limited to positive values
* Invocation metrics and all “count” style metrics (such as number of errors, number of customers, etc) will be made type NaturalNumber
* Response time and TPS metrics will remain as numbers but will be explicitly limited to positive values

Metric Changes:
* The `AvailabilityMetrics`, `PerformanceMetrics`, `AverageTPSMetrics` and `PeakTPSMetrics` models will be updated to include a separation between metrics for unauthenticated and authenticated metrics
* The ErrorMetrics model will be changed from a number to an object with the following:
    * Two fields containing objects named `authenticated` and `unauthenticated` to separate the errors for authenticated vs unauthenticated APIs
    * Each of these objects will contain a series of fields with the label of each field being a HTTP Status Code (e.g. 422, 500, etc) and the value being a number indicating the number of errors for the period
    * If a specific status code was not returned then a field for that code should be absent
    * Success codes (i.e. 2xx) should not be included

New Metrics:
A new `authorisations` field will be added to Get Metrics containing the following new metrics:
* `activeAuthorisationCount` - The number of active ongoing authorisations for a data holder at the time of the end point being invoked
* `newAuthorisationCount` - The number of new authorisations established during the period, reported on a daily basis.  This metric will be split between once-off and ongoing counts.
* `revokedAuthorisationCount` - The number of ongoing authorisations actively revoked during the period, reported on a daily basis.
* `amendedAuthorisationCount` - The number of ongoing authorisations amended during the period, reported on a daily basis.
* `expiredAuthorisationCount` - The number of ongoing authorisations expired during the period, reported on a daily basis.
* `abandonedConsentCount` - The number of initiated consent authorisation flows abandoned (for any reason) during the period, reported on a daily basis
* Each of the above counts will be split between `individual` and `nonIndividual` counts except for `abandonedConsentCount` which will be a single number 

In addition, language will be added to the standards to clarify that the Get Metrics API is intended to be called at the Data Holder Brand level (rather than the level of legal entity).

### V5 - Obligation Date Y23 #5 (11/03/2024)

The `PerformanceMetrics` model will be broken down into tiers similar to the `InvocationMetricsV2` model.  Each of the entries will however be an array of `RateString` values instead of a single value.  Each element of the array will represent a 1 hour period for the represented day with the 1st entry representing 12am-1am using the timezone used by the Data Holder.

A new `abandonmentsByStage` object will be added to `authorisations`.  This object will have the following entries:
* `preIdentification` - The number of authorisations that commenced with the data holder but the customer did not successfully identify their profile or user ID
* `preAuthentication` - The number of authorisations where the customer identified themselves but failed to provide a valid OTP or equivalent
* `preAccountSelection` - The number of authorisations where the customer successfully authenticated but abandoned the process before selecting accounts
* `preAuthorisation` - The number of authorisations where the customer successfully authenticated but abandoned the process before approving or rejecting the consent being requested
* `rejected` - The number of authorisations where the customer actively rejected the authorisation rather than abandoning the process
* `failedTokenExchange` - The number of authorisations that did not complete because token exchange with the ADR did not complete successfully

**Comment by perlboy** [comment]
A few questions as an initial gambit here, there is _definitely_ plenty of ambiguity which will make it quite difficult to commence implementation.

-  A proposed "end state" structure of V5 with an annotated V4 highlight (or vice versa) would be useful to understand the intended outcome.
- For HTTP Status Code
  - Some clarity will be required as to where it is expected this will be reported from. Currently many (most?) implementations have instrumented metric generation from the (micro)service itself, this may not be the http response experienced by the Data Recipient with respect to border control (which may simply timeout, be responsible for throwing 429s or return 200s with "Go Away") or load balancer routing behaviours (504 response codes). Indeed, some responses  may _never_ be observed by the service. Nearly all implementations will have almost zero visibility of status codes returned from DDoS protection providers. All of this means that deriving insights from this sort of data across holders is likely to be problematic.
  - Within underlying OAuth2 specifications there are multiple http status codes that are considered acceptable (ie. 400 vs. 422) driven by HTTP 1.0 vs. HTTP 1.1 aligned behaviours. This means this metric _won't_ be a consistent read across holders.
- The domains of the V4 and V5 authorisation values seem to be overlapping in many spaces and it's unclear if one is introduced and then replaced by the other while there seems to be some confusion around "consents" being inside authorisations. Additionally when considering amendments it's unclear what the abandonment stage would be, I can definitely see scenarios that are going to result in ambiguity and inquiry from the Regulator who have historically not understood the difference between an authentication, authorisation, consent, arrangement. Clear guidelines, ideally with a breakdown of scenarios aligned with the CX Guidelines would be highly beneficial.
- `RateString` is _really_ looking out of place now and I suspect is going to be more and more confusing as more industries are added. Is there any reason why a gradual transition to `PercentString` as a copy paste with the example removed wouldn't be appropriate?
- For `abandonmentsByStage.preIdentification` is this intended to be inclusive of requests which had PAR lodgement but for which consent flow never started (ie. the user-agent never turned up at the Holder)?
- For `abandonmentsByStage.preIdentification` is this intended to be inclusive of requests which are made _after_ the PAR request has expired (ie. a failure mode that renders no such identification step)?
- For `abandonmentsByStage.preAuthentication` is the customer considered to be identified _before or after_ the selection of their Consumer (ie. individual vs. non-individual consumer with the same user)?
- For `abandonmentsByStage.preAuthorisation` guidance will be required where and how in a flow involving joint accounts and possibly JAMS such a checkpoint is reached. Additionally, like below is this additive in nature (ie. _any_ checkpoint post authentication is included in `preAuthorisation`)?
- For `abandonmentsByStage.rejected` is this additive to other stages, ie. if a user never identifies and then clicks Cancel is this both a `rejected` _and_ a `preIdentification` abandonment? If not, at what point is `rejected` applicable?
- For `abandonmentsByStage. failedTokenExchange` this is _unlikely_ to be accurate as _many_ of the reasons for a failed token exchange are beyond consent (ie. expired refresh token, access token exchange issues), may never be seen (ie. code exchange by user-agent may result in an ADR never _actually_ asking for a token), may fail during processing for a variety of system related reasons that may impact metric collection. There's a special note here because observability by way of http status is almost impossible _by design_ because essentially all token failures are generic 400 errors.
- What is the expected migration plan? Many metric implementations have deliberately setup long term data structures in alignment with the existing metrics definitions. At a _minimum_ there would be a 7 day lag of probably inaccurate data to flush out old collection and replace it with new. While it is possible to maintain continuity this is additional scope with non-trivial engineering effort _purely_ to achieve a 7 day roll over.
 - Will it be acceptable to leap frog to V5 immediately?
  - Further on this, is it acceptable to assume that ACCC will be requesting `x-v`=5 by November and therefore a big bang cutover that resets counters to 0 at the new format will be reasonable?

**Comment by CDR-API-Stream** [comment]
Thanks @perlboy for the comprehensive feedback.  Much appreciated.  Feedback is below…

> * A proposed "end state" structure of V5 with an annotated V4 highlight (or vice versa) would be useful to understand the intended outcome.

I’ll add a proposed structure to the comment with the proposal

> * For HTTP Status Code
>   
>   * Some clarity will be required as to where it is expected this will be reported from. Currently many (most?) implementations have instrumented metric generation from the (micro)service itself, this may not be the http response experienced by the Data Recipient with respect to border control (which may simply timeout, be responsible for throwing 429s or return 200s with "Go Away") or load balancer routing behaviours (504 response codes). Indeed, some responses  may _never_ be observed by the service. Nearly all implementations will have almost zero visibility of status codes returned from DDoS protection providers. All of this means that deriving insights from this sort of data across holders is likely to be problematic.

We would probably best address this via guidance.  Nils recently updated all of the metrics guidance and merged it into a single article and we could extend that article as we go.  There is a potentially an endless set of clarifications for data holder specific issues that may arise and these would be too granular to include in the standards.  In general, however, we would align to the principle that errors should be reported as they are presented to the ADR.

>   * Within underlying OAuth2 specifications there are multiple http status codes that are considered acceptable (ie. 400 vs. 422) driven by HTTP 1.0 vs. HTTP 1.1 aligned behaviours. This means this metric _won't_ be a consistent read across holders.

That is understood.  There are already many implementation specific issues that make the metrics inconsistent across data holders.  The trend for any specific data holder should still be consistent over time, however.

> * The domains of the V4 and V5 authorisation values seem to be overlapping in many spaces and it's unclear if one is introduced and then replaced by the other while there seems to be some confusion around "consents" being inside authorisations. Additionally when considering amendments it's unclear what the abandonment stage would be, I can definitely see scenarios that are going to result in ambiguity and inquiry from the Regulator who have historically not understood the difference between an authentication, authorisation, consent, arrangement. Clear guidelines, ideally with a breakdown of scenarios aligned with the CX Guidelines would be highly beneficial.

We will look to provide this clarity by expanding guidance but there is probably improved clarity that can be added in the standards as well.  Specifically I will try to address all of the questions around exactly when each phase cuts over in the proposal.

Regarding the measurement of these calculations during amendment.  That isn’t an issue that has come up before and is possibly something we should address.  Do you think that we should separate metrics for amendments from new authorisations?

> * `RateString` is _really_ looking out of place now and I suspect is going to be more and more confusing as more industries are added. Is there any reason why a gradual transition to `PercentString` as a copy paste with the example removed wouldn't be appropriate?

I could argue that, technically, a RateString doesn’t mean interest rate but the mathematical contcept of a ratio of two measures over time.  That would be pretty pedantic on my part though as I'm sure most people simply assume the financial measure is implied.

Transitioning from RateString to PercentString is something that has been discussed before but has never seemed worth the effort.  If the timing is right to discuss it, however, maybe we should raise a maintenance CR and tackle it directly.  I’d rather not deal it with as a side effect of this consultation.

> * What is the expected migration plan? Many metric implementations have deliberately setup long term data structures in alignment with the existing metrics definitions. At a _minimum_ there would be a 7 day lag of probably inaccurate data to flush out old collection and replace it with new. While it is possible to maintain continuity this is additional scope with non-trivial engineering effort _purely_ to achieve a 7 day roll over.

The lead time for the proposed FDOs is intended to give space for a migration plan for each data holder.  In most cases, existing metrics are unaffected.  The type changes are really just API formatting and should not impact instrumentation and the new metrics do not have an existing set of data that will be changed.

The splitting of authenticated from unauthenticated metrics is the only real impact to an existing metric and the feedback has not indicated that this will be a problem as many implementations already have them separated.

Are there specific aspects of the proposal that will need further transition accommodation?

> * Will it be acceptable to leap frog to V5 immediately?

Yes.  The phasing is to accommodate data holders that may have to make significant changes to instrument their platforms.  If that isn’t a big deal then jumping to v5 should be acceptable.

> * Further on this, is it acceptable to assume that ACCC will be requesting `x-v`=5 by November and therefore a big bang cutover that resets counters to 0 at the new format will be reasonable?

Making the requests with `x-v=5` and `x-min-v=3` would be the goal allowing data holders to upgrade their systems at their own pace.  There is no resetting of counters anticipated, however.

On the resetting of counters, while metrics that were not previously recorded cannot be returned for periods for which they weren’t collected there should be no need to reset counters for any of the existing metrics.

**Comment by CDR-API-Stream** [comment]
Here are the proposed changes to the Non-Functional Requirements for further feedback.  These are candidate changes to be proposed to the Chair unless there is feedback indicating they should change:

## Non-Functional Requirement Changes

### Tiering of Traffic Thresholds

As there was consensus support for a tiered approach to traffic thresholds based on number of active authorisations the DSB is proposing amendments to the standards as outlined below.  These thresholds have been developed from the data that the DSB has been able to obtain regarding actual TPS and authorisation metrics for existing data holders.

The following statements in the standards in the `Traffic Thresholds` section will be amended:
> For secure traffic (both Customer Present and Unattended) the following traffic thresholds will apply:
> 
> * 300 TPS total across all consumers
>
> For Public traffic (i.e. traffic to unauthenticated end points) the following traffic thresholds will apply:
> 
> * 300 TPS total across all consumers (additive to secure traffic)

These statements will be replaced with: 
> For secure traffic (both Customer Present and Unattended) the following traffic thresholds will apply:
> 
> * For Data Holders with 0 to 2,000 active authorisations, 200 TPS total across all consumers
> * For Data Holders with 2,001 to 5,000 active authorisations, 300 TPS total across all consumers
> * For Data Holders with 5,001 to 10,000 active authorisations, 350 TPS total across all consumers
> * For Data Holders with 10,001 to 25,000 active authorisations, 400 TPS total across all consumers
> * For Data Holders with 25,001 to 50,000 active authorisations, 450 TPS total across all consumers
> * For Data Holders with more than 50,000 active authorisations, 500 TPS total across all consumers
>
> For Public traffic (i.e. traffic to unauthenticated end points) the following traffic thresholds will apply:
> 
> * 300 TPS total across all consumers (additive to secure traffic)

Note that this will be a reduction in expectation for the vast majority of existing Data Holders and will be an increase in expectation for a small number of the most active Data Holders.

It is proposed that these changes will be tied to a Future Dated Obligation of Obligation Date Y23 No. 5 (13/11/2023)

### NFRs for Low Velocity Data

To a differentiation for the calling by ADRs of low velocity data sets the following text will be added to the `Data Recipient Requirements` sub-section in the `Non-functional Requirements` section of the standards.

> ### Low Velocity Data Sets
> For endpoints that provide access to data that is low velocity (ie. the data does not change frequently) the Data Recipient is expected to cache the results of any data they receive and not request the same resource again until the data may reasonably have changed.
>
> For low velocity data sets, if the same data is requested repeatedly a Data Holder may reject subsequent requests for the same data during a specified period.
>
> Identified low velocity data sets are to be handled according to the following table noting that:
> * the `Velocity Time Period` is a continuous period of time in which calls beyond a specific threshold MAY be rejected by the Data Holder
> * the `Allowable Call Volume` is the threshold number of calls to the same resource for the same arrangement above which calls MAY be rejected by the Data Holder
> 
> | Data Set | Impacted Endpoints | Velocity Time Period | Allowable Call Volume|
> |-|-|-|-|
> | NMI Standing Data | Get Service Point Detail | 24 hours | 10 calls |
> | Energy Usage Data | Get Usage For Service Point, Get Bulk Usage, Get Usage For Specific Service Points| 24 hours | 10 calls |
> | DER Data |Get DER For Service Point, Get Bulk DER, Get DER For Specific Service Points| 24 hours | 10 calls |

As this change is really an expansion of the requirement that ADRs `minimise traffic with Data Holders` most ADRs should already be minimising highly cacheable data so no future dated obligation will be placed on this change.  Feedback on this aspect of the proposal is welcome.

**Comment by anzbankau** [comment]
> Here are the proposed changes to the Non-Functional Requirements for further feedback. These are candidate changes to be proposed to the Chair unless there is feedback indicating they should change:
> 
> ## Non-Functional Requirement Changes
> ### Tiering of Traffic Thresholds
> As there was consensus support for a tiered approach to traffic thresholds based on number of active authorisations the DSB is proposing amendments to the standards as outlined below. These thresholds have been developed from the data that the DSB has been able to obtain regarding actual TPS and authorisation metrics for existing data holders.
> 
> The following statements in the standards in the `Traffic Thresholds` section will be amended:
> 
> > For secure traffic (both Customer Present and Unattended) the following traffic thresholds will apply:
> > 
> > * 300 TPS total across all consumers
> > 
> > For Public traffic (i.e. traffic to unauthenticated end points) the following traffic thresholds will apply:
> > 
> > * 300 TPS total across all consumers (additive to secure traffic)
> 
> These statements will be replaced with:
> 
> > For secure traffic (both Customer Present and Unattended) the following traffic thresholds will apply:
> > 
> > * For Data Holders with 0 to 2,000 active authorisations, 200 TPS total across all consumers
> > * For Data Holders with 2,001 to 5,000 active authorisations, 300 TPS total across all consumers
> > * For Data Holders with 5,001 to 10,000 active authorisations, 350 TPS total across all consumers
> > * For Data Holders with 10,001 to 25,000 active authorisations, 400 TPS total across all consumers
> > * For Data Holders with 25,001 to 50,000 active authorisations, 450 TPS total across all consumers
> > * For Data Holders with more than 50,000 active authorisations, 500 TPS total across all consumers
> > 

We are supportive of tiering to allow lower thresholds for Data Holders with fewer users, however we do not agree that the maximum TPS levels should change at this time. 

Any uplift of TPS beyond 300TPS has a large impact on Data holders. Given this, the proposal to introduce new tiers of up to 500TPS should be performed through a dedicated Decision Proposal. This will allow Data Holders visibility of this impactful change, and time to assess implementation considerations.

**Comment by Origin-Rachel** [comment]
I think there needs to be a distinction between the type of consumer that these non-functional requirements apply to. I refer in particular to rules around response times for the "Get Bulk" APIs, which have no upper limit as to the number of accounts expected in the response. This means, for example, that the expected response time for "Get Bulk Billing" for a consumer with 1 account is the same for a consumer with 10, 50, or 100 accounts, when in practice increasing the number of accounts naturally results in an increased response time due to the amount of data requested. 

I suggest reviewing the practicality of applying the same response time to every CDR customer. I note that these requirements seem written predominantly for retail/mass market consumers. It is possible for business customers in energy to have over 100+ accounts.

**Comment by AusBanking-Christos** [comment]
Dear DSB,

In light of the discussion today with the DSB and our Members, who are seeking further opportunity to provide additional feedback, can we please request extending the consultation for another week to 19 May 2023.  

Kindest Regards,
Australian Banking Association

**Comment by cuctrangsb** [comment]
We also agree with @AusBanking-Christos and would like to request for an extension of the consultation period to 19 May 2023.

Kind regards,
Great Southern Bank

**Comment by CDR-API-Stream** [comment]
This consultation will be extended until the 19th May as requested.  The DSB would prefer not extend this consultation any further beyond this date.

We understand the need for modifications to NFRs to be an ongoing process and to be based on objective data.  It would appear that this may require changes the NFRs to be supported by a more specific, regular and engaged consultation process.

To that end we are planning a series of workshops specifically on NFRs for the ecosystem in late July or early August.  These workshops will be used to work with the community to create an ongoing consultation process for NFRs that works for everyone as well as to canvas the community about any issues and solutions related to the NFR standards that the community wishes to raise.

More details on these workshops will be announced in due course.

**Comment by johnAEMO** [comment]
Comment on the NFRs for low velocity data:
- The proposed allowable call volume of 10 will need to make allowances for instances when there are multiple pages - we have just seen calls with 3 service points for Usage data which have spanned 8 pages.  The presumption is that each page can be called up to 10 times?
- This NFR is required! - AEMO has been subject to hits requesting a different page of Usage data for the same 3 NMI request every 6 seconds over the last 3 days.  While we can reconfigure our infrastructure to allow for this type of request pattern, we would contend that this is atypical and allowance should not be made.

**Comment by perlboy** [comment]
The feedback for this DP is significant making it difficult to comment further. What I'll note here is that it appears to discuss both NFR and Metrics endpoints simultaneously when the reality is that they are two separate spheres. Essentially NFRs set the thresholds and are likely to be more structurally architecture components while Metrics report on them which is more of an engineering activity. 

I suggest a more focused pair of DPs is proposed so that feedback can more easily be targeted on the specific areas.

**Comment by AusBanking-Christos** [comment]
18 May 2023

Submitted on Thursday, 18 May 2023 via: [Consumer Data Standards Australia - GitHub site](https://github.com/ConsumerDataStandardsAustralia/standards/issues/288)

Dear [@CDR-API-Stream](https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCDR-API-Stream&data=05%7C01%7CJulian.Luton%40cba.com.au%7C7cc4f852657d43522a0108db34a6fb53%7Cdddffba06c174f3497483fa5e08cc366%7C0%7C0%7C638161665657899499%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=buKD3iX%2BsiygVan5b9MpufesDPB%2FSOlOzHxp%2ByD7BUk%3D&reserved=0)

**ABA Follow Up Response to Decision Proposal 288 – Non-Functional Requirements**

ABA welcomes the plan for a series of workshops on NFRs for the CDR ecosystem.  The proposed multistakeholder approach will lay the foundations for a shared and transparent capacity planning framework that balances the needs of all participants while ensuring appropriate customer outcomes.

The workshops will provide the opportunity for richer performance data to be assessed when setting NFR standards.  ABA member banks commit to working with the DSB ahead of the meeting to identify a consistent data set that will be the most useful contribution to the workshop process.

We welcome the opportunity to contribute toward the development of NFR standards that will result in a sustainable and predictable capacity planning model for all CDR participants.

**Comment by jimbasiq** [comment]
Basiq feedback on the Traffic Thresholds proposed amendment is we are generally supportive but still concerned with the upper boundary. A highest limit dictated in 

> For Data Holders with more than 50,000 active authorisations, 500 TPS total across all consumers

 seems low considering Basiq currently have considerably more than 50,000 screen scrape active authorisations with each of the major banks, some in the hundreds of thousands. We intend to move all of these connections from screen scrape to open banking CDR connections.

If CDR intends to move data sharing from screen scraping to CDR it needs to both support the existing load and provide some overhead. I don't believe the current proposal is doing this.

**Comment by AGL-CDR** [comment]
Thank you for the opportunity to provide feedback on this area of discussion.

AGL does not support the tiering of thresholds for TPS for energy. 

This is because:

- There is currently no evidence that energy Data Holders are failing to meet existing TPS obligations.
- There is currently no evidence that CDR consumers accessing energy Data Holders have adverse experiences because of limitations inherent in the TPS NFR’s.
- Using Active Authorisations as a predictive benchmark for TPS volumes has not been established for energy. These brackets are therefore not based on ‘real world’ observed TPS levels where it comes to energy-related traffic.
- Active Authorisations in energy may fluctuate substantially more than in the banking sector because of the Multisite concept where hundreds (or possibly thousands) of authorisations may be initiated in a 24 hour period. This may result in traffic volume spikes that are difficult to plan for and may cause a temporary spike above a given TPS threshold before returning to previous ‘normal’ volumes.
- Ramping up capacity to meet hypothetical volumes is unrealistic as it will result in costly infrastructure upgrades that arrive after such spikes occur and there’s currently no evidence that higher TPS volumes would be sustained after an initial spike subsides. AGL would then incur considerably higher infrastructure costs associated with lengthy periods of unutilised capacity.

AGL requests that these proposed changes are delayed and revisited (for energy) until such time that at least twelve months of real-world traffic volumes have been observed following Tranche 3 Large Retailer Go Live. (**November 2024**)

**Comment by commbankoss** [comment]
Regarding additional consent metrics, CBA suggests a similar outcome, i.e. improvements to the consumer experience for consent flow to reduce drop-off rates, could be achieved through a consultative approach with ADRs and DHs. Our recommendation is that a sample of relevant consent metrics be amalgamated by participants and provided to the DSB as input. This approach would more cost effective for the ecosystem, achieve a similar outcome and avoid regret spend if authentication and consent flows are matured to enable Action Initiation in the future.

**Comment by anzbankau** [comment]
In light of recent discussions, ANZ requests that the proposed tiering remains conditional upon the outcome of the forthcoming workshops. Given the complex nature of open banking systems, meeting the revised tiering is unlikely to be a simple scaling out exercise. The workshops must consider that data holders will require extensive capacity planning, design and implementation activities.

**Comment by NationalAustraliaBank** [comment]
NAB welcomes the plan for further workshops on NFRs. As the topic appears to be of great interest to CDR community, we recommend these workshops are scheduled sooner rather than later to maintain the positive momentum.

We also acknowledge DSB feedback regarding white label implementations and are keen to engage with DSB and any other interested participants to explore the topic in detail. With regards to the proposed Get Metrics future dated obligations, we request that they are pushed back by one release cycle, i.e. that the v4 FDO is aligned with Y24 #1 (11/03/2024) and v5 is aligned with Y24 #2 (13/05/2024).

**Comment by WestpacOpenBanking** [comment]
Westpac welcomes the opportunity to respond to the additional proposals added to DP288.


**Scaling NFRs for Large Data Holders**

A tiered approach by activity is an improvement to the current standard. Nevertheless, Westpac suggests that the proposal needs to be evidence-based prior to structuring the tiering levels and thresholds. Our evidence suggests that current activity in the ecosystem do not warrant the unusually high thresholds in the current proposal. We welcome the opportunity to discuss the TPS proposal in the planned workshops for July-August and we support earlier comments that this is not ready for presentation to the DSB chair.

Westpac notes that it is difficult to set a fair and adequate TPS level without context of the use-cases that the ecosystem is wanting to support; since some use-cases require more load than others. We suggest that focus should be on activity growth in the medium term future only. Handling of larger volumes can be revisited as the ecosystem matures with clearer pipeline of future use-cases and activity types flowing in the ecosystem. This would allow better allocation and direction of investment that is aligned to the Govt intention as announced in the recent Budget.

**Proposed changes to existing metrics**

Westpac is broadly supportive of the proposed changes to existing metrics.


**Proposed new metrics**

Westpac notes from various comments above that there may be various uses to the statistics around ‘abandonment by stage’ by different parties within the ecosystem (regulators, ADRs, DHs, incumbents, and prospects). We suggest the following improvements to increase the value of new metrics prior to implementing changes: 

- **Scope efficiently the purpose of the metrics prior to implementation**: A more prudent approach to obtaining insights into abandonment issues would be to apply a “lean start-up approach”; for example, to first collect data as a one-off basis from Data Holders (ie manually produced) and making the results available for the ecosystem to consume and analyse. A consultation after this exercise would give better clarity to all participants regarding their needs, and provide better input as to whether the type of metric is sufficient and relevant to their needs/uses. 

- **Increasing value and insights from metrics:**

     •  Authorisation and abandonment metrics need to be associated to the software product that the authorisations are for. 
     
     •	 To be able to increase customer adoption in the ecosystem, it is more meaningful to understand the driver of the customer’s decision; and deriving metrics from DHs presuppose the problem of customer abandonment relates to the stages of the authorisation. From a consumer viewpoint, the reason for customer abandonment on average is unlikely to correlate to technical issues encountered during DH’s stages of the authorisation flow (eg customer’s change of mind with using ADR software product, cold-feet with data-sharing, staff-testing a select component of the flow).
     
    •	Westpac supports ABA’s position that the DSB allows a negative CX journey path for a “standardised set of reasons” that can be presented to customers at the point of customer abandonment. Furthermore, in advance of the suggested CX changes, a lean low-cost approach could be to launch a standardised one-off survey monkey distributed to genuine ADR’s customers that did not complete data-sharing authorisation (ie exclude staff testing activities). 

Westpac also notes that there are many comments and questions around the definition of the metrics that needs to be discussed and resolved prior, to presentation to the DSB Chair. Considering the nature and size of the change vary depending on these definitions, it would be more appropriate to set the delivery timelines after the conclusion of the discussions or workshops. We ask that in light of the current backlog of standard changes, 9 months be provided as minimum to allow organisations to budget resource and deliver. The ecosystem cannot sustain ongoing urgent revisions to standards as we have recently experienced with FAPI 1.0

**Comment by JohnMillsEnergyAustralia** [comment]
Thank you for this opportunity to make a submission. EnergyAustralia submits the following:

With the energy sector only so recently going live the existing NFRs for us as a Data Holder remain untested by ADR usage seen to-date. Therefore the need to revise the NFRs so dramatically, and then to apply these to energy sector would be premature.

We are aligned with the AGL submission made on this topic that is reflective of the energy sector. 

It appears that a staged approach to retain the existing NFRs for Energy may well prove more suitable to support a nascent CDR sector like Energy.  This will avoid the risk of over-funding capacity.  A sectorial approach should be based on CDR usage statistics from the Energy sector so when it reaches the maturity of banking sector it would see it move to the next stage of NFRs.  It would then see more appropriate NFRs for more mature sectors and existing NFRs for new to CDR sectors like Energy for their first two years.  

Publication of metrics of overall usage remains of benefit.  However more detailed publication of NFR metrics on such small numbers of usage is not really presently of industry benefit (until the 2 year point) following their CDR implementation, and only if volumes increase. Such limited usage will skew the usage to potentially mis-represent any valid conclusions drawn.

Further. we specifically endorse the final paragraph from the AGL submission on AEMO performance that concludes with “AGL considers that it would be appropriate for AEMO to establish its own service desk arrangement for the resolution of tickets directly with ADRs and reduce administrative pressures on data holders to manage these issues.”

**Comment by ACCC-CDR** [comment]
Ecosystem metrics data exposed by participants via the GetMetrics API is published by the ACCC on the CDR Performance Dashboard. Several of the changes proposed here will result in breaking changes for the Performance Dashboard, including breaking continuity of historical data and the ability to compare metrics between versions, which is a mandatory requirement for us to operate and regulate the system. While breaking changes are sometimes necessary to advance the API and enrichen metrics data over time, the `averageTps` and `peakTps` overall values need to be maintained in the metrics payload to facilitate this continuity. As the CDR ecosystem grows, our experience shows that it is impractical for all participants to transition between API versions in unison. A carefully constructed transition plan will therefore be necessary to ensure CDR metrics data relating to all participants remains publicly available. Given that we will be responsible for making the necessary changes to the Performance Dashboard and already work closely with participants, we will take the lead on this transition. Retirement dates for v3 should not be set until we have developed a transition plan.

In reference to the JSON schemas for v4 and v5 as posted [in the above comment](https://github.com/consumerdatastandardsaustralia/standards/issues/288#issuecomment-1515771012), we provide the following feedback:

- If the plan is to introduce additional granularity into the metrics provided, we need to ensure that the overall values are retained for continuity of historical data.
-   As previously stated in our [comment above](https://github.com/consumerdatastandardsaustralia/standards/issues/288#issuecomment-1491310954) we would like to split the performance metric by performance tier (i.e. Unauthenticated, High Priority, Low Priority, Unattended etc.)
-   Providing an aggregated performance metric for all authenticated calls does not provide enough granularity to assess the stated NFR.

**Comment by CDR-API-Stream** [comment]
Thanks everyone for the final feedback.  This thread will now be locked and responses will be posted and a final decision created for submission to the chair.

Note that the final decision on TPS thresholds may take a few more days as the DSB have been offered additional data that may influence the specifics of the thresholds to be set.

**Comment by CDR-API-Stream** [comment]
Response to feedback on low latency data clusters:
* There is general consensus that this approach is appropriate
* There hasn't been any feedback seeking a change to the current proposal

In response to the question from AEMO: yes, the current proposal would allow for each page of usage history to be called up to 10 times per day.  If the threshold needs to be adjusted based on actual experience that can be consulted on in the future.  In the interim this should allow for calls being made every couple of minutes to be managed to protect core systems

**Comment by CDR-API-Stream** [comment]
Response to feedback on changes to the Get Metrics API:

**Additional Changes**
* As per the request from the ACCC, we will add in fields for the existing aggregated metrics for the purposes of trend continuity.  This will be in both v4 and v5 of the Get Metrics API
* As per the request from the ACCC, instead of splitting performance metrics between authenticated and unauthenticated we will these metrics by tier.  A metric for the aggregated NFR of 99.5% of all invocations within performance requirements will be retained.  This will be included in v5 of the Get Metrics API

**Implementation Considerations**
* In response to the request for a later implementation date for the Get Metrics changes we will propose retaining the existing date for v4 and pushing back the date for v5 by one milestone.  The reason for retaining the v4 date is so that tranche 3 energy retailers going liver at the end of this will be reporting data consistently in the new format from the beginning but also to align to the feedback received that it would be better to begin getting the improved data earlier rather than later
* A deprecation date for v3 of Get Metrics will not be set at this time.  The ACCC have indicated they will provide a schedule for their adoption of v4 and v5.  Once this is done a deprecation date for v3 will be able to be set.

**Responses to other feedback**

Some participants suggested other mechanisms for providing data rather than updating the Get Metrics API.  We have not modified our proposal in response to this feedback for two reasons:
1. We were specifically requested for improvements to the Get Metrics API by the regulator to assist in the ongoing management of the ecosystem and to reduce the cost of ad hoc reporting.  These changes are in line with that request
2. Reception of suggestions related to the voluntary provision of data is influenced by the lack of response to such requests for data by the DSB in the past.  It is not clear that a mechanism for voluntary provision of data would be effective

The suggestion that authorisation abandonment metrics should be aligned to software product has not be incorporated into the proposal.  Doing so would increase implementation costs but would also be of minimal value as the proposed metrics only come into play once software product process has successfully completed (ie. the customer has already accepted the proposed consent presented by the ADR).  The metrics are only representative of the data holder screens which are common across all software products.  We may consider this feedback in the future if the concept of data recipient metrics is introduced to CDR.

It was suggested that energy retailers should not be required to provide metrics until the 2 year point of implementation.  This is not consistent with the requirements applied to the banking sector in the past but, more importantly, is not a question being considered in this consultation so no action is being taken on this feedback.

**Comment by CDR-API-Stream** [comment]
The DSB has received data from multiple banks related to the number of active authorisations and TPS levels.  Each of the banks that provided detailed data ask that it be kept confidential so it will not be published but it does give a stronger evidence foundation for setting the TPS tiering.

As a result of this data it would appear that our initial proposal (which was based only on number of customers) was far too aggressive and should be altered significantly.

The new proposed tiering for site wide authenticated peak TPS will therefore be:
* For Data Holders with 0 to 10,000 active authorisations, 150 peak TPS total across all consumers
* For Data Holders with 10,001 to 20,000 active authorisations, 200 peak TPS total across all consumers
* For Data Holders with 20,001 to 30,000 active authorisations, 250 peak TPS total across all consumers
* For Data Holders with 30,001 to 40,000 active authorisations, 300 peak TPS total across all consumers
* For Data Holders with 40,001 to 50,000 active authorisations, 350 peak TPS total across all consumers
* For Data Holders with 50,001 to 60,000 active authorisations, 400 peak TPS total across all consumers
* For Data Holders with more than 60,000 active authorisations, 450 peak TPS total across all consumers

Note that the implications of this tiering strategy is as follows:
* No existing data holder will have any immediate obligation increase
* The majority of data holders will get a significant reduction in their site wide obligation (most data holders fit inside the lowest two tiers currently)
* Energy retailers will get a significant reduction in obligation.  The tranche 3 retailers going live at the end of the year will probably remain in the bottom tier for the medium term
* For ADRs that are planning to migrate screen scarping customers there should now be enough headroom for interacting with the larger banks as the number of consents being established grows

Responses to specific feedback by participants is as follows:
* There is consensus support for a tiering approach to site wide peak TPS thresholds.
* A number of data holders requested that change be made right now in advance of the proposed workshops later this year.  Considering the length of time already invested in consulting on this issue and the existing problems being articulated by the ADR community this is not the preferred approach.
* There was a request to exclude energy retailers from these changes.  The NFRs in the standards are deliberately cross-sectoral and the intent is for this to remain the case.  The reasons for this exclusion seems to be related to ensuring the obligation on energy retailers is not increased.  As the proposed changes results in a significant *reduction* in obligation for energy retailers (including the tranche 3 retailers) these issues seem to be adequately addressed.
* There was a suggestion to change the NFRs for energy C&I customers specifically.  This feedback may be important to address but does not seem aligned to the specific issue being consulted on.  This feedback will be noted and raised again during the NFR workshops to be held later this year.
*

**Comment by CDR-API-Stream** [comment]
The Data Standards Chair has approved Decision 288 attached below:
[Decision 288 - Non-Functional Requirements Revision.pdf](https://github.com/ConsumerDataStandardsAustralia/standards/files/11811640/Decision.288.-.Non-Functional.Requirements.Revision.pdf)

It is intended that this decision will be published in the standards in v1.25.0 in the next two weeks.

**Comment by nils-work** [comment]
[Standards](https://consumerdatastandardsaustralia.github.io/standards/index.html) version 1.25.0 has now been published, incorporating the changes detailed in the decision [above](https://github.com/ConsumerDataStandardsAustralia/standards/issues/288#issuecomment-1599976783).
